File: search-in-rotated-sorted-array/search-in-rotated-sorted-array.cpp
class Solution {
    int b_s2(vector<int>nums,int ind,int target){
        int lo=ind,hi=nums.size()-1;
        while(lo<=hi){
            int mid=lo+(hi-lo)/2;
            if(nums[mid]==target)
                return mid;
            else if(nums[mid]>target)
                hi=mid-1;
            else
                lo=mid+1;
        }
        return -1;
    }
    int b_s1(vector<int>nums,int ind, int target){
        int lo=0,hi=ind;
        while(lo<=hi){
            int mid=lo+(hi-lo)/2;
            if(nums[mid]==target)
                return mid;
            else if(nums[mid]>target)
                hi=mid-1;
            else
                lo=mid+1;
        }
        return -1;
    }
    public:
    int search(vector<int>& nums, int target) {
        int lo=0,hi=nums.size()-1;
        while(lo<hi){
            int mid=lo+(hi-lo)/2;
            if(nums[mid]>nums[hi])
                lo=mid+1;
            else
                hi=mid;
        }
        return b_s1(nums,hi-1,target)==-1?b_s2(nums,hi,target):b_s1(nums,hi-1,target);
        // return hi;
    }
};

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        def binary_search(arr):
            l = 0
            h = len(arr)-1
            while(l<=h):
                mid = l+(h-l)//2
                if arr[mid]==target:
                    return mid
                elif arr[mid]>target:
                    h = mid-1
                else:
                    l = mid + 1
            return -1
        low = 0
        high = len(nums)-1
        while(low<high):
            mid = low+(high-low)//2
            if nums[mid]>nums[high]:
                low = mid+1
            else:
                high = mid
        left_search = binary_search(nums[:low])
        right_search = binary_search(nums[low:])
        return low + right_search if right_search != -1 else left_search